Add structure viewer to the analysis report

SarahAlidoost opened the pull request:
You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md) and that you comply with the following criteria:

- [ ] You have sticked to Python. Please talk to us before adding other programming languages to HADDOCK3
- [ ] ~~Your PR is about CNS~~
- [x] Your code is well documented: proper docstrings and explanatory comments for those tricky parts
- [x] You structured the code into small functions as much as possible. You can use classes if there is a (state) purpose
- [x] Your code follows our coding style
- [ ] ~~You wrote tests for the new code~~
- [x] `tox` tests pass. *Run `tox` command inside the repository folder*
- [x] `-test.cfg` examples execute without errors. *Inside `examples/` run `python run_tests.py -b`*
- [x] PR does not add any dependencies, unless permission granted by the HADDOCK team
- [x] PR does not break licensing
- [ ] ~~Your PR is about writing documentation for already existing code :fire:~~
- [ ] ~~Your PR is about writing tests for already existing code :godmode:~~

---

This PR adds the [NGL](http://nglviewer.org/ngl/gallery/) molecule viewer to the `report.html` file generated by `analyse` command. It also refactors tables in the report and now uses some HTML and CSS syntax. :see_no_evil:
Here is the list of further improvements which can be developed in other PRs: 
- enbale toggling clusters between plots and tables
- enable downloading all pdb files, and plots in one-go
- enable select 4 best structures
- be able to sort inside the table using cluster Rank
- add some of the ngl tools e.g. rotate, show water

SarahAlidoost commented:
@mgiulini and @sverhoeven  can you please review this PR? Thanks.

SarahAlidoost commented:
need to be fixed:
- NGL bg color --> white
- use cluster rank instead of id --> tables and plots

sverhoeven commented on src/haddock/libs/libplots.py:
```diff
@@ -633,22 +636,114 @@ def find_best_struct(ss_file, number_of_struct=10):
+            var stage;
+
+            document.addEventListener("DOMContentLoaded", function () {
+                // Setup to load data from rawgit
+                NGL.DatasourceRegistry.add(
+                    "data",
+                    new NGL.StaticDatasource(
+                        "//cdn.rawgit.com/arose/ngl/v2.1.0/data/"
+                        )
+                );
```
I think this code does nothing. As we are loading files from the local http server. Can it be removed?

```suggestion
```

sverhoeven commented on src/haddock/libs/libplots.py:
```diff
@@ -706,36 +815,125 @@ def clt_table_handler(clt_file, ss_file):
+        border-collapse: collapse;
+        width: 100%;
+        font-size: 16px;
+        height: 500px;
```
When more scores are added then you can no longer see the best structures. You will need to scroll down to see them, which a user might not know.

Can the height be made dependent on the number of rows? For example use `12rem` as their are 1 header + 11 score rows + 1 best structure row.

SarahAlidoost replied:
Do you mean no vertical scroll in the table? then, I set the `height=auto`.  The table will automatically adjust its height based on the content. Is this what you meant? :thinking:

sverhoeven commented on src/haddock/libs/libplots.py:
```diff
@@ -633,22 +636,115 @@ def find_best_struct(ss_file, number_of_struct=10):
+        dl_name = dl_name + suffix
+
+        # Generate html code
+        html_code = "<p>"
```
The p-tag adds margin above and below making it look different then the other cells. 

Can you use `span` tag instead so each cell has same height?

SarahAlidoost replied:
fixed!

sverhoeven reviewed (approved):
Works for me. I am able to download and view the structures.

I have some inline suggestions which are mainly on cosmetics. So could also be done later.

SarahAlidoost commented on src/haddock/libs/libplots.py:
```diff
@@ -633,22 +636,114 @@ def find_best_struct(ss_file, number_of_struct=10):
+            <dialog id="structureViewerDialog">
+                <div id="viewport" style="width:800px; height:600px;"></div>
+                <form>
+                    <button value="cancel" formmethod="dialog">X</button>
```
Done!

SarahAlidoost commented on src/haddock/libs/libplots.py:
```diff
@@ -633,22 +636,114 @@ def find_best_struct(ss_file, number_of_struct=10):
+    """)
+    ngl_dialog = ("""
+        <body>
+            <dialog id="structureViewerDialog">
```
now, the viewer shows the file name with the download link.

SarahAlidoost commented on src/haddock/clis/cli_analyse.py:
```diff
@@ -425,12 +425,20 @@ def main(run_dir, modules, top_cluster, format, scale, **kwargs):
+        report_file = Path(outdir, directory, "report.html")
+        log.info(f"View the results in {report_file}")
+        info_msg = ("To view structures or download the structure files, "
+                    f"in the {outdir} directory and in a terminal run the "
```
added!

SarahAlidoost commented:
@mgiulini could you please review this PR?

SarahAlidoost commented:
> Works for me. I am able to download and view the structures.
> 
> I have some inline suggestions which are mainly on cosmetics. So could also be done later.

Thanks for reviewing and good comments :+1:

mgiulini commented:
> @mgiulini could you please review this PR?

sure, I will do it first thing tomorrow!

mgiulini commented on src/haddock/clis/cli_analyse.py:
```diff
@@ -301,8 +301,8 @@ def analyse_step(step, run_dir, capri_dict, target_path, top_cluster, format, sc
         ss_fname = Path(run_dir, f"{step}/capri_ss.tsv")
-        clt_fname = Path(run_dir, f"{step}/capri_clt.tsv")
         shutil.copy(ss_fname, target_path)
+        clt_fname = Path(run_dir, f"{step}/capri_clt.tsv")
```
what is this change? probably related to some space..try to avoid introducing such changes, as they make the history messy

mgiulini commented on src/haddock/libs/libplots.py:
```diff
@@ -706,36 +832,129 @@ def clt_table_handler(clt_file, ss_file):
 
 
-def report_generator(boxes, scatters, table, step):
+def _generate_html_body(figures):
+    body = "<body>"
+    include_plotlyjs = "cdn"
+    for figure in figures:
+        if isinstance(figure, str):  # tables
+            inner_html = f'''<div class="table">{figure}</div>'''
+        else:  # plots
+            inner_html = figure.to_html(
+                full_html=False, include_plotlyjs=include_plotlyjs
+                )
+            include_plotlyjs = False
+        body += "<br>"  # add a break between tables and plots
+        body += inner_html
+    body += _ngl_viewer()
+    body += "</body>"
+    return body
```
could you add a small docstring to this function? explaining what is figures (array of plots, I suppose)

SarahAlidoost replied:
added!

mgiulini commented on src/haddock/libs/libplots.py:
```diff
@@ -706,36 +832,129 @@ def clt_table_handler(clt_file, ss_file):
+    return [table]
+
+
+def _css_styles_for_report():
```
can you add a small explanation/docstring?

SarahAlidoost replied:
added!

mgiulini commented on src/haddock/libs/libplots.py:
```diff
@@ -706,36 +832,129 @@ def clt_table_handler(clt_file, ss_file):
+    return f'<style>{custom_css}</style>'
+
+
+def _generate_html_report(step, figures):
```
docstring (see above)

SarahAlidoost replied:
added!

mgiulini commented on src/haddock/libs/libplots.py:
```diff
@@ -706,36 +832,129 @@ def clt_table_handler(clt_file, ss_file):
+    return html_report
+
+
+def _generate_html_head(step):
```
docstring missing (see above)

SarahAlidoost replied:
added!

mgiulini commented on src/haddock/libs/libplots.py:
```diff
@@ -623,32 +631,133 @@ def find_best_struct(ss_file, number_of_struct=10):
+            if rn == "Unclustered":
+                dl_name = f'unclustered_model{str_number}'
+            else:
+                dl_name = f'cluster{rn:02d}_model{str_number}'
```
very minor, but why is this renumbering necessary? why not keeping the standard `cluster_N_model_M.pdb` nomenclature? I don't like seeing zeros in the file names eheh

SarahAlidoost replied:
fixed!

mgiulini reviewed (approved):
nice PR! it works nicely on a set of runs I analysed.

I just added a couple of comments about the introduction of docstrings and naming convention.

mgiulini commented:
I noticed a serious problem with the PR: with the previous implementation, the bar plots were built so as each cluster until rank 10, and subsequent clusters were lumped together in a single bar (the eleventh bar). As an example consider this plot
<img width="541" alt="image" src="https://github.com/haddocking/haddock3/assets/54807167/af7ded33-3cd2-43ea-822d-008cd9221ac2">


Now the clusters with rank > 10 are superimposed on the eleventh bar, making the plot quite ugly:
<img width="541" alt="image" src="https://github.com/haddocking/haddock3/assets/54807167/9823e0cd-91d2-4974-bd9d-ea874108ebbb">

can you please check?

SarahAlidoost commented:
> I noticed a serious problem with the PR: with the previous implementation, the bar plots were built so as each cluster is shown until rank 10, and subsequent clusters were lumped together in a single bar (the eleventh bar). As an example consider this plot <img alt="image" width="541" src="https://user-images.githubusercontent.com/54807167/242568573-af7ded33-3cd2-43ea-822d-008cd9221ac2.png">
> 
> Now the clusters with rank > 10 are superimposed on the eleventh bar, making the plot quite ugly: <img alt="image" width="541" src="https://user-images.githubusercontent.com/54807167/242568064-9823e0cd-91d2-4974-bd9d-ea874108ebbb.png">
> 
> can you please check?

Sure, I will check it. Can you tell me which example did you use to generate the plots?

rvhonorato commented:
This is a great addition, do we still need anything before this is merged?

SarahAlidoost commented:
> This is a great addition, do we still need anything before this is merged?

thanks, yes, I want to address the comments by Marco before merging it :blush:

mgiulini commented:
> Sure, I will check it. Can you tell me which example did you use to generate the plots?

Sure, I used a modified version of the docking-protein-protein-test.cfg in examples/docking-protein-protein:

```
# ====================================================================
# Protein-protein docking example with NMR-derived ambiguous interaction restraints

# directory in which the scoring will be done
run_dir = "run1-test"
ncores = 40

# postprocess caprieval folders with haddock3-analyse
postprocess = true
mode = "local"
# CNS executable path - optional
# cns_exec = "path/to/bin/cns"

# molecules to be docked
molecules =  [
    "data/e2aP_1F3G.pdb",
    "data/hpr_ensemble.pdb"
    ]

# ====================================================================
# Parameters for each stage are defined below, prefer full paths
# ====================================================================
[topoaa]
#tolerance = 20
autohis = false
[topoaa.mol1]
nhisd = 0
nhise = 1
hise_1 = 75
[topoaa.mol2]
nhisd = 1
hisd_1 = 76
nhise = 1
hise_1 = 15

[rigidbody]
tolerance = 20
ambig_fname = "data/e2a-hpr_air.tbl"
sampling = 200

[caprieval]
#tolerance = 20
reference_fname = "data/e2a-hpr_1GGR.pdb"

[clustfcc]

[caprieval]
#tolerance = 20
reference_fname = "data/e2a-hpr_1GGR.pdb"

[seletopclusts]

[caprieval]
#tolerance = 20
reference_fname = "data/e2a-hpr_1GGR.pdb"
```

SarahAlidoost commented:
@mgiulini thanks for checking this thoroughly. The problem in the box plot is fixed now. In addition to this fix, there are some new changes:
-  In this pull request, the legends of the plots (both scatters and box plots) are now cluster ranks instead of cluster id, as discussed in the previous meeting. I noticed that the legend is not sorted. The order is also fixed now. 
- I changed the `px_colors.qualitative.Alphabet` to `px_colors.qualitative.Dark24`. The reason is that `qualitative.Alphabet` has a very light color that might be picked by plots, see number 8 [here](https://plotly.com/python/discrete-color/).

SarahAlidoost commented:
@mgiulini and @rvhonorato I think the failing test is not related to the changes in this PR. Otherwise, please let me know.

rvhonorato commented:
Its failing because of some CI oversight, addressed it in #491 - good to merge if you greenlight @mgiulini

mgiulini commented:
hey @SarahAlidoost , I checked the new code, it works perfectly, well done! I'll merge it!
